File: widgetssdk/src/main/java/com/glia/widgets/view/head/ChatHeadLayout.java
package com.glia.widgets.view.head;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Pair;
import androidx.core.view.ViewCompat;

import com.glia.widgets.R;
import com.glia.widgets.UiTheme;
import com.glia.widgets.core.configuration.GliaSdkConfiguration;
import com.glia.widgets.di.Dependencies;
import com.glia.widgets.helper.Utils;
import com.glia.widgets.view.ViewHelpers;
import com.google.android.material.theme.overlay.MaterialThemeOverlay;

public class ChatHeadLayout extends FrameLayout implements ChatHeadLayoutContract.View {
    private ChatHeadView chatHeadView;
    private ChatHeadLayoutContract.Controller controller;

    private NavigationCallback navigationCallback;
    private OnChatHeadClickedListener chatHeadClickedListener;

    private UiTheme uiTheme;

    private boolean isChatView = false;

    public ChatHeadLayout(@NonNull Context context) {
        this(context, null);
    }

    public ChatHeadLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.gliaChatStyle);
    }

    public ChatHeadLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, R.style.Application_Glia_Chat);
    }

    public ChatHeadLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(
                MaterialThemeOverlay.wrap(
                        context,
                        attrs,
                        defStyleAttr,
                        defStyleRes),
                attrs,
                defStyleAttr,
                defStyleRes
        );
        init(attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public void showOperatorImage(String operatorImgUrl) {
        chatHeadView.showOperatorImage(operatorImgUrl);
    }

    @Override
    public void showUnreadMessageCount(int count) {
        chatHeadView.showUnreadMessageCount(count);
    }

    @Override
    public void showPlaceholder() {
        chatHeadView.showPlaceholder();
    }

    @Override
    public void showQueueing() {
        chatHeadView.showQueueing();
    }

    @Override
    public void showOnHold() {
        chatHeadView.showOnHold();
    }

    @Override
    public void hideOnHold() {
        chatHeadView.hideOnHold();
    }

    @Override
    public void navigateToChat() {
        if (navigationCallback != null) {
            navigationCallback.onNavigateToChat();
        } else {
            chatHeadView.navigateToChat();
        }
    }

    @Override
    public void navigateToCall() {
        if (navigationCallback != null) {
            navigationCallback.onNavigateToCall();
        } else {
            chatHeadView.navigateToCall();
        }
    }

    @Override
    public boolean isInChatView() {
        return isChatView;
    }

    @Override
    public void show() {
        post(() -> setVisibility(VISIBLE));
    }

    @Override
    public void hide() {
        post(() -> setVisibility(GONE));
    }

    @Override
    public void setController(ChatHeadLayoutContract.Controller controller) {
        this.controller = controller;
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        float floatingViewX = w - getChatHeadSize() - getChatHeadMargin();
        float floatingViewY = h / 10f * 8f;
        chatHeadView.setX(floatingViewX);
        chatHeadView.setY(floatingViewY);
        chatHeadView.invalidate();
        super.onSizeChanged(w, h, oldw, oldh);
    }

    @Override
    protected void onDetachedFromWindow() {
        controller.onDestroy();
        super.onDetachedFromWindow();
    }

    /**
     * Method for the integrator to override if they want to do custom logic when the chat head is
     * clicked.
     *
     * @param listener
     */
    public void setOnChatHeadClickedListener(OnChatHeadClickedListener listener) {
        this.chatHeadClickedListener = listener;
    }

    /**
     * Method that allows integrator to override navigation on click with using own paths
     * <p>
     * if set to null default navigation is restored
     *
     * @param callback
     */
    public void setNavigationCallback(NavigationCallback callback) {
        this.navigationCallback = callback;
    }

    public void setConfiguration(GliaSdkConfiguration configuration) {
        this.chatHeadView.updateConfiguration(uiTheme, configuration);
    }

    public void setIsChatView(boolean value) {
        this.isChatView = value;
    }

    private void init(@Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        setVisibility(GONE);
        initConfigurations();
        initViews();
        setupViewActions();
        readTypedArray(attrs, defStyleAttr, defStyleRes);
        setController(Dependencies.getControllerFactory().getChatHeadLayoutController());
        this.controller.setView(this);
    }

    private void initConfigurations() {
        setClickable(false);
        setFocusable(false);
        ViewCompat.setElevation(this, 100.0f);
    }

    private void initViews() {
        View view = View.inflate(getContext(), R.layout.chat_head_layout, this);
        chatHeadView = view.findViewById(R.id.chat_head_view);
    }

    @SuppressLint("ClickableViewAccessibility")
    private void setupViewActions() {
        chatHeadView.setOnTouchListener(
                new ViewHelpers.ChatHeadOnTouchListener(
                        this::getChatHeadViewPosition,
                        this::onChatHeadDragged,
                        this::onChatHeadClicked
                )
        );
    }

    private void readTypedArray(AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        @SuppressLint("CustomViewStyleable") TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.GliaView, defStyleAttr, defStyleRes);
        setBuildTimeTheme(Utils.getThemeFromTypedArray(typedArray, getContext()));
        typedArray.recycle();
    }

    private void setBuildTimeTheme(UiTheme theme) {
        this.uiTheme = theme;
        this.chatHeadView.updateConfiguration(uiTheme, null);
    }

    private void onChatHeadDragged(float x, float y) {
        chatHeadView.setX(x);
        chatHeadView.setY(y);
        chatHeadView.invalidate();
    }

    private Pair<Integer, Integer> getChatHeadViewPosition() {
        return new Pair<>(
                Float.valueOf(chatHeadView.getX()).intValue(),
                Float.valueOf(chatHeadView.getY()).intValue()
        );
    }

    private void onChatHeadClicked(View v) {
        if (chatHeadClickedListener != null) {
            chatHeadClickedListener.onClicked(null);
        } else {
            controller.onChatHeadClicked();
        }
    }

    private float getChatHeadSize() {
        return getResources().getDimension(R.dimen.glia_chat_head_size);
    }

    private float getChatHeadMargin() {
        return Utils.pxFromDp(this.getContext(), 16);
    }

    public interface OnChatHeadClickedListener {
        void onClicked(GliaSdkConfiguration chatHeadInput);
    }

    public interface NavigationCallback {
        void onNavigateToChat();

        void onNavigateToCall();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/view/head/ChatHeadLayout.kt
package com.glia.widgets.view.head

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.PointF
import android.util.AttributeSet
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updateLayoutParams
import com.glia.widgets.R
import com.glia.widgets.databinding.ChatHeadLayoutBinding
import com.glia.widgets.di.Dependencies
import com.glia.widgets.helper.layoutInflater
import com.glia.widgets.helper.wrapWithGliaTheme
import com.glia.widgets.internal.chathead.BubblePosition
import com.glia.widgets.internal.chathead.ChatBubbleContract
import com.glia.widgets.internal.chathead.domain.ResolveChatHeadNavigationUseCase.Destinations
import com.glia.widgets.view.SimpleTouchListener
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

internal class ChatHeadLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(
    context.wrapWithGliaTheme(),
    attrs,
    defStyleAttr
) {
    private val chatBubbleController: ChatBubbleContract.Controller by lazy { Dependencies.controllerFactory.chatBubbleController }

    private var scope: CoroutineScope? = null

    private var navigationCallback: NavigationCallback? = null

    private val _chatHeadViewPosition: PointF
        get() = PointF(chatHeadView.x, chatHeadView.y)

    private val chatHeadSize: Float by lazy { resources.getDimension(R.dimen.glia_chat_head_size) }
    private val chatHeadBottomRightMargin: Float by lazy { resources.getDimension(R.dimen.glia_chat_head_content_padding) }
    private val chatHeadTopLeftMargin: Float by lazy { resources.getDimension(R.dimen.glia_small) }

    private val binding: ChatHeadLayoutBinding by lazy {
        ChatHeadLayoutBinding.inflate(layoutInflater, this)
    }

    private val chatHeadView: ChatHeadView by lazy { binding.chatHeadView }

    init {
        initialize()
        z = 100f // Make sure chat head is on top of other views
    }

    private fun navigateToChat() {
        navigationCallback?.onNavigateToChat() ?: chatHeadView.navigateToChat()
    }

    private fun navigateToCall() {
        navigationCallback?.onNavigateToCall() ?: chatHeadView.navigateToCall()
    }

    /**
     * Positions the bubble once the bounds are known: at the place the visitor dragged it to — shared
     * with the overlay bubble and mapped into this layout's own range, which also carries it across
     * rotations — or at the default corner if it was never dragged.
     */
    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        val xMax = w - chatHeadSize - chatHeadBottomRightMargin
        val yMax = h - chatHeadSize - chatHeadBottomRightMargin
        val position = chatBubbleController.bubblePosition

        chatHeadView.x = position?.xWithin(chatHeadTopLeftMargin, xMax) ?: xMax
        chatHeadView.y = position?.yWithin(chatHeadTopLeftMargin, yMax) ?: (h / 10f * 8f)
        chatHeadView.invalidate()
        super.onSizeChanged(w, h, oldw, oldh)
    }

    private fun saveBubblePosition() {
        chatBubbleController.bubblePosition = BubblePosition.within(
            x = chatHeadView.x,
            y = chatHeadView.y,
            xMin = chatHeadTopLeftMargin,
            xMax = width - chatHeadSize - chatHeadBottomRightMargin,
            yMin = chatHeadTopLeftMargin,
            yMax = height - chatHeadSize - chatHeadBottomRightMargin
        )
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        scope?.cancel()
        scope = null
    }

    /**
     * Method that allows integrator to override navigation on click with using own paths
     *
     *
     * if set to null default navigation is restored
     *
     * @param callback
     */
    fun setNavigationCallback(callback: NavigationCallback) {
        navigationCallback = callback
    }

    private fun initialize() {
        initConfigurations()
        setupViewActions()
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        // Being attached is the visibility mechanism: the watcher only attaches this layout while the
        // controller says the bubble renders in the app, and detaches it otherwise.
        visibility = VISIBLE

        CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate).also { scope = it }.launch {
            chatBubbleController.uiState.collect(chatHeadView::render)
        }

        // Request to apply insets so that we can adjust the chat head position for edge-to-edge
        ViewCompat.requestApplyInsets(this)

        // Listen for inset changes
        setupInsetsListener()
    }

    private fun setupInsetsListener() {
        ViewCompat.setOnApplyWindowInsetsListener(this) { view, insets ->
            // Instead of using the `insets` parameter directly (which might have been
            // consumed by a parent), we can get the original, unaltered insets
            // for the entire window using `getRootWindowInsets`. This is more reliable
            // when you don't control the parent Activity.
            val rootInsets = ViewCompat.getRootWindowInsets(view)
            if (rootInsets == null) {
                // If we can't get root insets, return the original ones.
                return@setOnApplyWindowInsetsListener insets
            }

            // Get the insets for the system bars (status bar and navigation bar).
            val systemBarInsets = rootInsets.getInsets(WindowInsetsCompat.Type.systemBars())

            // Apply the insets as margin to this view.
            // This will ensure that the chat head does not go beneath the system bars.
            view.updateLayoutParams<MarginLayoutParams> {
                setMargins(
                    systemBarInsets.left,
                    systemBarInsets.top,
                    systemBarInsets.right,
                    systemBarInsets.bottom
                )
            }
            // Return the original insets so that other views in the hierarchy
            // can also process them. Do not consume them here.
            insets
        }
    }

    private fun initConfigurations() {
        isClickable = false
        isFocusable = false
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun setupViewActions() {
        chatHeadView.setOnTouchListener(
            SimpleTouchListener(
                retrieveInitialCoordinates = { _chatHeadViewPosition },
                onMove = { x, y -> onChatHeadDragged(x, y) },
                onRelease = {
                    saveBubblePosition()
                    ChatHeadLogger.logPositionChanged()
                }
            )
        )
        chatHeadView.setOnClickListener { onChatHeadClicked() }
    }

    private fun onChatHeadDragged(x: Float, y: Float) {
        // Make sure the chat head doesn't go off the screen
        chatHeadView.x = x.coerceIn(chatHeadTopLeftMargin, width - chatHeadSize - chatHeadBottomRightMargin)
        // Make sure the chat head doesn't go off the screen
        chatHeadView.y = y.coerceIn(chatHeadTopLeftMargin, height - chatHeadSize - chatHeadBottomRightMargin)
        chatHeadView.invalidate()
    }

    private fun onChatHeadClicked() {
        when (chatBubbleController.onBubbleTapped()) {
            Destinations.CALL_VIEW -> navigateToCall()
            Destinations.CHAT_VIEW -> navigateToChat()
        }
    }

    fun removeSelf() {
        (parent as? ViewGroup)?.removeView(this)
    }

    interface NavigationCallback {
        fun onNavigateToChat()
        fun onNavigateToCall()
    }
}
